feat(elevation): move elevation rules into theme stylesheets

benelliott opened the pull request:
Move all usages of elevation mixins out of component stylesheets and into their respective themes.
This allows the elevation color to be changed based on the theme. Add new `mat-theme-elevation`
and `mat-theme-overridable-elevation` shorthand mixins to apply the current theme's elevation
color. Replace all usages of `mat-elevation` mixins with themed equivalents. Add `elevation`
property to the default `foreground` theme palettes.

Closes #11343

BREAKING CHANGE:

Because `mat-elevation` usages have been moved out of component stylesheets, users who have
not invoked a theme mixin will not see any elevation shadows on Material components.
However, users that have created a custom theme which lacks the `elevation` property will
still see the default black shadows.

Additionally, users who want to use themed elevations in their custom components can create
their own shorthand mixin:

```
@import '~angular/material/theming';

$myTheme: ...

@mixin my-elevation($zValue) {
  @include mat-theme-elevation($zValue, $myTheme);
}

```

and then invoke `angular-material-theme` with the `$myTheme` variable.

benelliott commented:
Example of using this feature in a custom theme - setting `elevation` to `violet`:

![violet-elevation](https://user-images.githubusercontent.com/4996462/40073356-84230e12-586e-11e8-973a-b08838a80f43.png)

jelbourn commented:
@benelliott are you saying this is only breaking when a theme is missing? If that's the case, it wouldn't be considered breaking because a theme has always been required.

benelliott commented:
@jelbourn The situations I can think of are:

1.  The user uses a built-in theme - these themes have the `foreground.elevation` property set to `black` so this is the same as before
2. The user uses a custom theme - the `mat-theme-elevation` mixin defaults the color to `black` if the theme map returns `null` so this is the same as before
3. The user uses no theme - in this case, there will be no shadow

So yes I think so, unless there is a situation I'm not thinking of!

jelbourn commented on src/lib/core/style/_elevation.scss:
```diff
@@ -151,6 +151,14 @@ $_mat-elevation-prefix: 'mat-elevation-z';
               #{map-get(_get-ambient-map($color, $opacity), $zValue)};
 }
 
+@mixin mat-theme-elevation($zValue, $theme, $opacity: $mat-elevation-opacity) {
```
The new mixins need descriptions similar to the existing ones

jelbourn commented on src/lib/core/style/_elevation.scss:
```diff
@@ -151,6 +151,14 @@ $_mat-elevation-prefix: 'mat-elevation-z';
 
+@mixin mat-theme-elevation($zValue, $theme, $opacity: $mat-elevation-opacity) {
+  $foreground: map-get($theme, foreground);
+  $elevation-color: map-get($foreground, elevation);
```
Typically the background and foreground palettes are fixed regardless of theme; how would you go about changing the color of the shadow?

benelliott replied:
I took the position that wanting to change the elevation color was a fairly advanced use case, so it would be admissible for it to be changed by providing `angular-material-theme` with a fully custom theme map (as opposed to using the `mat-light-theme` and `mat-dark-theme` functions). This is what I do in a project where I want full control over the colours used in the app, however it has just occurred to me that it may not be part of the public API(?).

An alternative solution could be to provide it as an optional fourth argument to the `mat-light-theme` and `mat-dark-theme` functions (with default value `black`), and then to convert the `$mat-light-theme-foreground` and `$mat-dark-theme-foreground` variables into functions that take the elevation color as their only argument.

jelbourn replied:
@josephperrott and I took a closer look at this, and found that we're going to have some cases in Google where it will be more common to want to customize the shadow color based on the theme colors, so we probably want to make it more ergonomic to customize the shadow based on that.

We'd also want to avoid adding a lot of excess css to the theme so most people wouldn't end up using. I'm still thinking about a good way to approach it.

benelliott replied:
@jelbourn Perhaps the additional parameter to the `theme` functions could be used instead?

An alternative way to avoid adding CSS to the themes could be to remove all internal usages of the mixins, and only to rely on the `.mat-elevation-z?` classes for applying elevation to Material components. This approach would actually reduce the amount of CSS overall, as only the output of

```
 @for $zValue from 0 through 24 {
    .#{$_mat-elevation-prefix}#{$zValue} {
      @include mat-theme-elevation($zValue, $theme);
    }
  }
```

would be rendered (component stylesheets would be smaller as they no longer include the box shadow rules).

jelbourn replied:
I like the idea of using the common set of css classes for everything. We would have to make sure they're still easily overridable, though. 

I guess what this PR is really about is being able to customize the _base_ shadow color, which makes sense to be in the foreground palette. If we want to set a shadow based on another palette, I don't see a way around explicitly passing the color in the component's theme file.

jelbourn commented:
@benelliott also, could you DM me your email address on twitter?

jelbourn commented:
@crisbeto do you have any opinions on this? I would lean towards just making the elevation part of the theme solely because it's a simpler change

crisbeto commented:
Keeping it in the theme sounds good. My only nit is that, IMO, the new mixins should be made private (prefixed with an underscore) since I don't see them being particularly useful for consumers.

benelliott commented:
Rebased and added suggestion from @crisbeto.

benelliott commented:
@crisbeto Is there anything else you would like me to change before this can be merged?

jelbourn commented:
I'm running a presubmit of this PR against Google apps to make sure it doesn't cause any breakages

benelliott commented:
Cool, thanks.

mmalerba commented:
@jelbourn what happened with the presubmit on this one, should we mark lgtm, merge-ready?

jelbourn commented:
Presubmit was black across the board, need to update some build files for another pass
(I went through one round of this already but ended u getting merge conflicts and had to restart)

benelliott commented:
Hi @jelbourn, I've just rebased to resolve the conflicts. Is black good or bad for a presubmit?

jelbourn commented:
Black is broken; it just needed the build configuration to be updated since some file dependencies changed

mmalerba commented:
Marking merge-ready so the next caretaker knows to presubmit it.

benelliott commented:
Resolved conflicts.

benelliott commented:
Resolved conflicts.

mmalerba commented:
@benelliott I think I managed to sort out what's wrong in google3, once you rebase I should be able to confirm and we can finally merge this

benelliott commented:
@mmalerba Rebased, thanks. Would you be able to describe roughly what the problem was, out of curiosity?

mmalerba commented:
There was a component internally using the `mat-menu-base` mixin. Removing the parameter caused it to break

benelliott commented:
Ah I see. Is that a public API? If so I might need to list it as a breaking change in the commit message.

mmalerba commented:
@jelbourn do we consider that part of the public API?

Update on the presubmit - changing those usages fixed the majority of issues, there are still ~10 screenshot tests failing showing differences in elevation of cards and buttons. I suspect it's a specificity issue, they probably have some overrides that need to be made more specific

jelbourn commented:
Only the theming mixins are part of the public API, since we don't publish the rest to npm.

jelbourn commented:
Oh, and things like `mat-core()` and `cdk-a11y`

mmalerba commented:
Another update: Internally we have some components based on these but with additional styles. I'll need to make a similar update to these internal components and send it out to the various teams using them before we can merge this one

benelliott commented:
Thanks for the updates!

mmalerba commented:
@benelliott I've changed all of the internal components to only use elevation in their theme files, if you rebase I'll take another stab at getting this in

benelliott commented:
@mmalerba Nice - rebased. Not sure why ngbot didn't ping me about the conflicts this time round.

mmalerba commented on src/lib/button/_button-base.scss:
```diff
@@ -86,10 +85,6 @@ $mat-mini-fab-padding: 8px !default;
-  }
-
   &[disabled] {
     box-shadow: none;
```
I think this needs to move to the theme too, I'm seeing screenshots with disabled buttons that have shadows (this is basically the equivalent of `mat-elevation(0)`)

mmalerba replied:
card has a similar one that should go into the theme file:
```scss
&.mat-card-flat {
  box-shadow: none;
}
```

benelliott commented:
Hi @mmalerba, I've addressed your comments + rebased + given the flat card and disabled buttons a quick once-over on the demo app.

I decided to use elevation 0 instead of `box-shadow: none` to preserve the level of abstraction. Let me know if you want me to change that.

benelliott commented:
🎉🎉🎉
